lib/SILOptimizer/Mandatory/TFDevicePartition.cpp

```diff
@@ -828,14 +826,57 @@ class DevicePartitionerImpl
                         unsigned operIdx) {
     auto opValue = inst->getOperand(operIdx);
 
-    // BB args are replicated on all devices, so no transfer is needed.
-    if (isa<SILArgument>(opValue)) return;
     // Undefs will be handled later.
     if (isa<SILUndef>(opValue)) return;
+    if (isa<SILArgument>(opValue)) {
+      // BB args that are not func input args are replicated on all devices, so
```

Comment:
Also not for this patch, but:

Now that graph_op has taken over, I think we should reconsider the marking pass / device partitioning divide, and go to the "bitvector" marking approach, which each value gets marked with the set of devices it has to be live on.  BB arguments really shouldn't be replicated on all devices.  With TPUs for example, all the weights that are being trained would end up being copied between the CPU and TPU on each step.

Reply:
Yes this has been on my TODO list.